### PR TITLE
Make hasUniqueRepresentation false for slices

### LINKED ISSUE #7330: std.AutoHashMap allows string keys
I think this is a bug, feel free to close if this is intentional.  The `autoHash` function contains this check, which seems like it is meant to disallow string keys in AutoHashMap:
```zig
if (comptime meta.trait.isSlice(Key)) {
    comptime assert(@hasDecl(std, "StringHashMap")); // detect when the following message needs updated
    const extra_help = if (Key == []const u8)
        " Consider std.StringHashMap for hashing the contents of []const u8."
    else
        "";

    @compileError("std.auto_hash.autoHash does not allow slices (here " ++ @typeName(Key) ++
        ") because the intent is unclear. Consider using std.auto_hash.hash or providing your own hash function instead." ++
        extra_help);
    }
}
```
However, before running this code, it does this check:
```zig
if (comptime trait.hasUniqueRepresentation(K)) {
    return Wyhash.hash(0, std.mem.asBytes(&key));
} else {
     var hasher = Wyhash.init(0);
     autoHash(&hasher, key);
     return hasher.final();
}
```
`trait.hasUniqueRepresentation(..)` hits this case for slices:
```zig
.Pointer => return true,
```
So `autoHash` never gets compiled for the case where the key is a string, and the compile error does not trigger.

### PR DESCRIPTION
**Pull request by indocomsoft**
Fixes #7330 

## Questions
- Should I remove the corresponding slice check from `autoHash`?
- Is there any way to test that `@compileError` is emitted?

### CODE REVIEW

**Review by Vexu — CHANGES_REQUESTED** [review]
The proper fix is to change `meta.trait.hasUniqueRepresentation` to return `.Pointer => |info| return info.size != .Slice,` for pointers.

> * Is there any way to test that `@compileError` is emitted?

Not yet: #513.

**Comment by indocomsoft** [comment]
@Vexu fixed according to your comment.

I've also moved the check for `[]const u8` to AutoHashMap as @Sahnvour said in #7330. Let me know if I should move this change to another PR instead.

File: lib/std/hash/auto_hash.zig [hunk 1]
```diff
@@ -169,21 +169,31 @@ pub fn hash(hasher: anytype, key: anytype, comptime strat: HashStrategy) void {
     }
 }
 
+fn typeContainsSlice(comptime K: type) bool {
+    comptime {
+        if (meta.trait.isSlice(K)) {
+            return true;
+        }
+        if (meta.trait.is(.Struct)(K)) {
```
Comment (Vexu) [inline lib/std/hash/auto_hash.zig:177]: This check should also be done on unions with an enum tag since we also hash their fields.

**Comment by indocomsoft** [comment]
@Vexu I've just added code to handle unions. I wonder if for `hasUniqueRepresentation` I should iterate through each union field and ensure that `@sizeOf(FieldType) == @sizeOf(T)` instead

File: lib/std/meta/trait.zig [hunk 2]
```diff
@@ -490,6 +490,22 @@ pub fn hasUniqueRepresentation(comptime T: type) bool {
...
+            // Only packed and extern unions have guaranteed in-memory representation
+            switch (info.layout) {
+                else => return false,
+
+                .Extern, .Packed => {
+                    inline for (info.fields) |field| {
+                        const FieldType = field.field_type;
+                        if (comptime !hasUniqueRepresentation(FieldType)) return false;
+                    }
+                    return true;
```
Comment (Vexu) [inline lib/std/meta/trait.zig:503]: Unless all fields have the same bit size the union will contain padding bits for some fields, I think it is ok to just leave this as a false.

Comment (indocomsoft) [reply lib/std/meta/trait.zig]: Gotcha, I just did this.

File: lib/std/meta/trait.zig [hunk 3]
```diff
@@ -529,10 +530,53 @@ test "std.meta.trait.hasUniqueRepresentation" {
...
+    };
+
+    testing.expect(hasUniqueRepresentation(TestUnion1));
+
+    const TestUnion2 = extern union {
+        a: u32,
+        b: u16,
+    };
+
+    testing.expect(hasUniqueRepresentation(TestUnion2));
```
Comment (Vexu) [inline lib/std/meta/trait.zig:557]: `!hasUniqueRepresentation` on both of these.

**Review by Vexu — APPROVED** [review]
Thanks for making the extra effort to do this properly.

File: lib/std/meta/trait.zig (hunk 3 above)
Comment (indocomsoft) [reply lib/std/meta/trait.zig]: Ah, I missed these. Fixed.

Thank you for helping to review my PR!

**Pull request merged**